Add global python startup script support in addition to per-user scripts

alexbruy opened the pull request:
We already have support for per-user Python startup scripts. This adds support for global startup script for all users. Script file should be called `global_startup.py` and placed in the QGIS Python directory, e.g. `$QGIS_INSTALL_PATH/python/global_startup.py`.

Some use cases: restore/update settings, add connections etc.

alexbruy committed:
Add global python startup script support in addition to per-user scripts (fix #6297)

alexbruy commented:
@NathanW2 assigning you, as you developed support for user startup scripts

NathanW2 commented:
Generally I'm not a fan of users putting stuff into the install folder.
Good feature though.  Is there a shared location we can use?

I will see where a good place on Windows is if others know the others.

On Wed, 27 Apr 2016 2:53 am Alexander Bruy <notifications@github.com> wrote:

> Assigned #3035 <https://github.com/qgis/QGIS/pull/3035> to @NathanW2
> <https://github.com/NathanW2>.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/3035#event-641772659>
>

wonder-sk commented:
Just curious... what is the use case for this functionality?

NathanW2 commented:
Could be handy for a Enterprise type setup too bootstrap stuff in

On Wed, 27 Apr 2016 1:49 pm Martin Dobias <notifications@github.com> wrote:

> Just curious... what is the use case for this functionality?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/3035#issuecomment-214963101>
>

alexbruy commented:
@wonder-sk as @NathanW2 said this can be used in enterprise environments to setup some stuff, another possible use case is resetting settings on each run (can be useful in computer classes)

@NathanW2 I don't know about any shared location which can be used in all systems. On Windows we maybe can use `C:\Users\AllUsers` directory, and for Linux use  `/etc`.

NathanW2 commented:
Qt should let us get those in a cross platform way I would say.

On Wed, 27 Apr 2016 4:49 pm Alexander Bruy <notifications@github.com> wrote:

> @wonder-sk <https://github.com/wonder-sk> as @NathanW2
> <https://github.com/NathanW2> said this can be used in enterprise
> environments to setup some stuff, another possible use case is resetting
> settings on each run (can be useful in computer classes)
>
> @NathanW2 <https://github.com/NathanW2> I don't know about any shared
> location which can be used in all systems. On Windows we maybe can use
> C:\Users\AllUsers directory, and for Linux use /etc.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/3035#issuecomment-214986942>
>

yjacolin commented:
@alexbruy it is a new feature and should be added in the documentation, isn't? If so could you add FEATURE tags in a commit (or create a ticket in the QGIS doc repository with information, example)? Thanks!

alexbruy commented:
Checked Qt docs, the most close things are QDesktopService for Qt4 and QStandardPaths for Qt5. But seems QDesktopService returns only user-specific paths

m-kuhn commented:
> Checked Qt docs, the most close things are QDesktopService for Qt4 and QStandardPaths for Qt5. But seems QDesktopService returns only user-specific paths

If QStandardPaths returns the proper paths, just wrap it in an `#if QT_VERSION >= 0x050000` block

alexbruy commented:
We need to find solution for Qt4 which is main target platform for now

m-kuhn commented:
Sure, just saying that if Qt4 doesn't offer it I wouldn't mind using something like

```
#if QT_VERSION >= 0x050000
  locations = QStandardPaths::standardLocations(QStandardPaths::AppDataLocation);
#else
  locations = QStringList() << "/usr/local/share" << "/usr/share";
#endif
```

m-kuhn commented:
Or probably better

```
#if QT_VERSION >= 0x050000
  location = QStandardPaths::locate(QStandardPaths::AppDataLocation, "global_startup.py");
#else
  location = "/usr/local/share/QGIS/global_startup.py";
#endif
```

wonder-sk commented:
What is wrong with using ```pythonPath()``` as proposed in the patch now?

NathanW2 commented:
Putting stuff in the install folder feels bad to me, also breaks when you do updates, etc.

wonder-sk commented:
But from what I understand, such ```global_startup.py``` file would be inserted there during the custom installation, so that seems OK. Updates would again be handled by the same installer provider, so I can't see a problem there...

NathanW2 commented:
Ok. All good with me if others agree.

On Wed, 4 May 2016 4:47 pm Martin Dobias <notifications@github.com> wrote:

> But from what I understand, such global_startup.py file would be inserted
> there during the custom installation, so that seems OK. Updates would again
> be handled by the same installer provider, so I can't see a problem there...
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/3035#issuecomment-216759788>
>

m-kuhn commented:
I also think a clear separation between application provided files and organisation provided files should be maintained in the folder structure. Am I getting something wrong here?

(Note, my above proposal for file locations may not be in line with this, mainly caused by my lack of FHS)

wonder-sk commented:
> I also think a clear separation between application provided files and organisation provided files should be maintained in the folder structure. Am I getting something wrong here?

I do not have a strong opinion, but it just seems easier to mix in custom files into the existing folder structure when making a custom installer. Is there any advantage in separating them?

m-kuhn commented:
> Is there any advantage in separating them?

You don't risk losing your self-maintained files when clearing or updating files which have been provided by upstream (and therefore are easy to obtain again). I haven't seen a software shipping config files in /bin.

NathanW2 commented:
I agree that was my main point. Also makes testing easier as you can just
point all installs there

On Wed, 4 May 2016 5:05 pm Matthias Kuhn <notifications@github.com> wrote:

> Is there any advantage in separating them?
>
> You don't risk losing your self-maintained files when clearing or updating
> files which have been provided by upstream (and therefore are easy to
> obtain again). I haven't seen a software shipping config files in /bin.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/3035#issuecomment-216762224>
>

m-kuhn commented:
I actually really like [`QStandardPaths::locate`](http://doc.qt.io/qt-5/qstandardpaths.html#locate) or [`QStandardPaths::locateAll`](http://doc.qt.io/qt-5/qstandardpaths.html#locateAll) since they have multiple locations, some of which only writable to sys-admins, some of them in the user folder. Thinking about a university setup where I want my QGIS to load this amazing python script I wrote but the sys-admin don't want to give me write access to system folders.

m-kuhn commented:
```
>>> QStandardPaths.standardLocations(QStandardPaths.AppDataLocation)
['/home/mku/.local/share/qgis', '/usr/local/share/qgis', '/usr/share/qgis']
```

Ship files to `/usr/share/qgis` if you have a custom installer

Put files into `/usr/local/share/qgis` as sysadmin

Put files into `/home/mku/.local/share/qgis` per user

m-kuhn commented:
For reference, this issue will be fixed: http://hub.qgis.org/issues/6297

alexbruy commented:
@m-kuhn wrote
> Thinking about a university setup where I want my QGIS to load this amazing python script I wrote but the sys-admin don't want to give me write access to system folders.

But you already can put **your** script in the your user folder and it will be executed on every QGIS start.

> Ship files to `/usr/share/qgis` if you have a custom installer
> Put files into `/usr/local/share/qgis` as sysadmin
> Put files into `/home/mku/.local/share/qgis` per user

So it returns different paths depending on user privileges. This is not what we need here, as global startup script location should be the same for all users regardless of their privileges.

m-kuhn commented:
> But you already can put your script in the your user folder and it will be executed on every QGIS start

True, but wouldn't a unified way for user and system setup be nice instead of inventing the wheel (and maintaining the code) twice? (That does not mean that I would disable the current user script location immediately).

> So it returns different paths depending on user privileges.

I don't think so. The way I understand it it either returns the most specific occurrence of the file (`locate()`) or all occurrences of the file (`locateAll()`). Unrelated to privileges.

alexbruy commented:
> True, but wouldn't a unified way for user and system setup be nice instead of inventing the wheel (and maintaining the code) twice? (That does not mean that I would disable the current user script location immediately).

What do you mean by "unified way": same location for user and global scripts or using `QStandardPaths` class to get location of these scripts?

Anyway, seems for Qt4 (which is our major platform now) we should hardcode script location. At least I can't find a way to get it programatically with Qt.

m-kuhn commented:
> What do you mean by "unified way": same location for user and global scripts or using QStandardPaths class to get location of these scripts?

I was thinking of the same filename which will be searched with `QStandardPaths::locateAll` and every file found will be loaded. System file first, user file last.
In a `#if QT_VERSION` check with the `#else` containing hardcoded locations (one single system path but compatible to what QStandardPaths returns). What do you think?

alexbruy commented:
> I was thinking of the same filename which will be searched with QStandardPaths::locateAll and every file found will be loaded. System file first, user file last.

Hmm, sounds reasonable, but IMHO in this case we should use same approach for other QGIS stuff like user-defined expressions, plugins, styles etc. It won't be good to have some user stuff in `$HOME/.qgis2` and other in `$APPDATA/qgis`.

> In a #if QT_VERSION check with the #else containing hardcoded locations (one single system path but compatible to what QStandardPaths returns). What do you think?

I'm fine with this. For Windows 7+ we can use something like `C:\ProgramData\qgis`, for Linux `/usr/local/share/qgis`. Do have any experience with Mac? I suspect we can use same directory as in Linux for it, but not sure.

m-kuhn commented:
```
>>> QtCore.QStandardPaths.standardLocations(QtCore.QStandardPaths.AppDataLocation)
['/Users/mkuhn/Library/Application Support', '/Library/Application Support', '/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/Resources/Python.app/Contents/Resources/']
```

m-kuhn commented:
Not sure how this is influenced by homebrew, @dakcarto may know more

NathanW2 commented:
Status on this?  I don't mind where it goes as  long as others are happy and it's in the docs.  It's a admin feature anyway so it's doesn't need to be in a super friendly place and would be scripted by most admins anyway e.g via an login script.

alexbruy commented:
@m-kuhn, @NathanW2 pull-request updated and now uses QStandardPath approach for searching global script.

m-kuhn commented:
Sweet, that is

On linux: "~/.local/share/<APPNAME>", "/usr/local/share/<APPNAME>", "/usr/share/<APPNAME>"
On windows: "C:/Users/<USER>/AppData/Roaming/<APPNAME>", "C:/ProgramData/<APPNAME>", "<APPDIR>", "<APPDIR>/data"
?

alexbruy commented:
I have no windows around right now (will try to test tomorrow), but on linux it looks for startup script in
`$HOME/.local/share`, `/usr/local/share` and `/usr/share`.

In future we can modify it to have user script in `$HOME/.local/share` too and maybe move other stuff here.

NathanW2 commented:
@alexbruy that looks good to me.  Thanks for making that change.

m-kuhn commented:
Can we drop the existing user script in favour of the new one? The new one should be able to replace the old one if it uses `QStandardPath::locateAll()`.

alexbruy commented:
> Can we drop the existing user script in favour of the new one? The new one should be able to replace the old one if it uses QStandardPath::locateAll().

Yes, we can. I will update PR to handle multiple startup files.

alexbruy committed:
drop old user startup script in favour of the new system

User script should be placed into $HOME/.local/share/QGIS.
Additionally global scripts from /usr/local/share and
/usr/share will be loaded if present

alexbruy commented:
PR updated.

m-kuhn commented:
Very nice, thanks!
This could possibly be added to other things as well (svg icons, expressions, processing scripts...)

NathanW2 commented:
Yes that would be very nice.  We have been doing a fair bit of enterprise-style QGIS rollouts and that kind of thing helps in those environments.

alexbruy commented:
Agree, adopting this approach for other stuff would be good.

NathanW2 commented:
Merge this as a start, we can do the rest later

On Thu, 3 Nov 2016 8:27 pm Alexander Bruy <notifications@github.com> wrote:

> Agree, adopting this approach for other stuff would be good.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/3035#issuecomment-258107457>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3Mfg4PigILYChMoYRdnVvDlOuBgXks5q6bcagaJpZM4IQJvP>
> .
>
